File: src/policy-change-digest/digest_demo/workflow.py
# SPDX-License-Identifier: Apache-2.0
"""One durable session per revision pair, with explicit transcript recovery."""

import csv
import difflib
import fcntl
import io
import json
import os
import re
from pathlib import Path

from munarium_client.models import TurnResult

from .fixtures import digest, encode, read, save, verify, write
from .server import query_client


def compare(inputs, case):
    manifest = verify(inputs)
    if case not in manifest["cases"]:
        raise ValueError("Unknown revision pair")
    folder = Path(inputs) / case
    before = (folder / "before.md").read_text()
    after = (folder / "after.md").read_text() if (folder / "after.md").exists() else None
    old_rule = re.search(r"^Rule: (.+)$", before, re.M).group(1)
    new_rule = re.search(r"^Rule: (.+)$", after, re.M).group(1) if after else None
    return {
        "case": case,
        **manifest["cases"][case],
        "status": "insufficient_evidence"
        if after is None
        else "unchanged"
        if old_rule == new_rule
        else "candidate_impacts",
        "old_rule": old_rule,
        "new_rule": new_rule,
        "source_hashes": {
            name: sha for name, sha in manifest["files"].items() if name.startswith(case + "/")
        },
        "diff": "".join(
            difflib.unified_diff(
                before.splitlines(True),
                (after or "").splitlines(True),
                fromfile=case + "/r1",
                tofile=case + "/r2",
            )
        ),
    }


def validate(result, comparison, config):
    errors = []
    try:
        completion = result.completion
        if not completion:
            raise ValueError("No completion")
        if completion.model != config["model"] or completion.provider != (
            "ollama" if config["provider"] == "fixture" else config["provider"]
        ):
            errors.append("Unexpected completion identity")
        if completion.verification and completion.verification.violations:
            errors.append("Unresolved verification violations")
        answer = json.loads(completion.text.strip().removeprefix("```json\n").removesuffix("\n```"))
        if (
            set(answer) != {"status", "explanation", "candidates"}
            or not isinstance(answer["explanation"], str)
            or not 1 <= len(answer["explanation"]) <= 3000
        ):
            raise ValueError("Unexpected digest schema")
        if answer["status"] != comparison["status"]:
            errors.append(
                "Answer disagrees with revision availability or deterministic rule comparison"
            )
        if not isinstance(answer["candidates"], list) or len(answer["candidates"]) > 10:
            raise ValueError("Invalid candidate list")
        if answer["status"] != "candidate_impacts" and answer["candidates"]:
            errors.append("Unsupported candidates without a known change")
        if answer["status"] == "candidate_impacts" and not answer["candidates"]:
            errors.append("No supported candidate identified")
        labels = {f"{h.collection}/{h.chunk_id}": h for h in result.hits}
        seen = set()
        for candidate in answer["candidates"]:
            if set(candidate) != {"checklist_id", "explanation", "citations"}:
                raise ValueError("Unexpected candidate schema")
            if (
                not isinstance(candidate["explanation"], str)
                or not 1 <= len(candidate["explanation"]) <= 2000
            ):
                raise ValueError("Invalid candidate explanation")
            ident = candidate["checklist_id"]
            if not isinstance(ident, str) or ident in seen:
                raise ValueError("Duplicate or invalid checklist identity")
            seen.add(ident)
            citations = candidate["citations"]
            if (
                not isinstance(citations, list)
                or not citations
                or not all(isinstance(c, str) and c in labels for c in citations)
            ):
                errors.append("Citations do not resolve to retrieved chunks")
                continue
            hits = [labels[c] for c in citations]
            filenames = {h.source_path.rsplit("/", 1)[-1] for h in hits}
            if not {"before.md", "after.md", "checklist.md"}.issubset(filenames):
                errors.append("Candidate lacks both revisions and the affected checklist")
            if not any(f"Checklist ID: {ident}\n" in h.text for h in hits):
                errors.append("Checklist ID is absent from its cited evidence")
        for hit in result.hits:
            name = comparison["case"] + "/" + hit.source_path.rsplit("/", 1)[-1]
            if (
                comparison["source_hashes"].get(name) != hit.source_content_hash
                or comparison["case"] not in hit.source_path
            ):
                errors.append("Source revision hash or scope mismatch")
        return answer, errors
    except (ValueError, TypeError, KeyError, AttributeError):
        return None, ["Completion is not a supported cited digest"]


def export(work, journal, result, config):
    comparison = journal["comparison"]
    answer, errors = validate(result, comparison, config)
    raw = result.model_dump(mode="json")
    if journal.get("recovered"):
        raw.pop("skipped", None)
    packet = {
        "comparison": comparison,
        "analysis_status": "unverified" if errors else "validated_draft",
        "answer": answer,
        "errors": errors,
        "evidence": raw,
        "exhaustive": False,
        "runbook": journal["runbook"],
        "revision_runbooks": config["revision_runbooks"][comparison["case"]],
        "input_hash": journal["input_hash"],
    }
    save(work / "digest.json", packet)
    text = f"# Candidate policy impacts: {comparison['case']}\n\nAnalysis: {packet['analysis_status']}. Coverage: retrieved candidates only; not exhaustive.\nRevision pair: {comparison['before']} → {comparison['after']}\n\n"
    text += (
        answer["explanation"] if answer and not errors else "Requires review: " + "; ".join(errors)
    ) + "\n\n"
    if answer and not errors:
        for candidate in answer["candidates"]:
            text += f"- Candidate {candidate['checklist_id']}: {candidate['explanation']}\n"
            text += "  Citations: " + ", ".join(candidate["citations"]) + "\n"
    text += (
        "\n## Exact textual diff\n\n```diff\n"
        + comparison["diff"]
        + "```\n\n## Source revisions\n\n"
    )
    text += "".join(
        f"- {name}: SHA-256 {sha}\n" for name, sha in comparison["source_hashes"].items()
    )
    write(work / "digest.md", text.encode())
    output = io.StringIO(newline="")
    writer = csv.writer(output)
    writer.writerow(
        ["case", "before", "after", "candidate_checklist", "analysis_status", "coverage"]
    )
    for candidate in answer["candidates"] if answer and not errors else []:
        writer.writerow(
            [
                comparison["case"],
                "r1",
                "r2",
                candidate["checklist_id"],
                packet["analysis_status"],
                "not exhaustive",
            ]
        )
    write(work / "digest.csv", output.getvalue().encode())
    return packet


def process(inputs, credentials, work, case, recover=False, crash=False):
    work = Path(work)
    work.mkdir(parents=True, exist_ok=True)
    with (work / ".lock").open("a+b") as lease:
        fcntl.flock(lease, fcntl.LOCK_EX | fcntl.LOCK_NB)
        comparison = compare(inputs, case)
        api, config = query_client(Path(credentials))
        with api:
            if (
                config["fixture_revision"]
                != digest((Path(inputs) / "manifest.json").read_bytes())[:12]
            ):
                raise ValueError("Credential fixture revision mismatch")
            runbook = config["runbooks"][case]
            fingerprint = digest(
                encode(
                    {
                        "comparison": comparison,
                        "runbook": runbook,
                        "model": config["model"],
                        "provider": config["provider"],
                        "uid": config["uid"],
                    }
                )
            )
            path = work / "journal.json"
            if path.exists():
                journal = read(path)
                if journal["input_hash"] != fingerprint:
                    raise ValueError(
                        "Changed revision pair or configuration requires a new work directory"
                    )
                if journal.get("result"):
                    return export(
                        work, journal, TurnResult.model_validate(journal["result"]), config
                    )
                if not recover or not journal.get("session_id"):
                    raise RuntimeError(
                        "Uncertain turn requires explicit transcript inspection; no replay"
                    )
                transcript = api.sessions.get(journal["session_id"])
                if transcript.uid != config["uid"] or transcript.runbook_ref != runbook:
                    raise ValueError("Transcript identity mismatch")
                turns = [
                    t for t in transcript.turns if t.query == journal["query"] and t.completion
                ]
                if len(turns) != 1:
                    raise RuntimeError("No uniquely recoverable completion; turn remains uncertain")
                turn = turns[0]
                completion = dict(turn.completion)
                completion["was_override"] = completion["resolved"]["was_override"]
                result = TurnResult(
                    session_id=transcript.session_id,
                    ordinal=turn.ordinal,
                    collections_searched=turn.collections_searched,
                    hits=turn.hits or [],
                    envelopes=turn.envelope or [],
                    completion=completion,
                )
                journal["recovered"] = True
            else:
                query = (
                    "Compare the policy revisions and identify candidate downstream checklist impacts. REVISION_COMPARISON="
                    + json.dumps(comparison, ensure_ascii=False, sort_keys=True)
                )
                journal = {
                    "state": "creating_session",
                    "input_hash": fingerprint,
                    "comparison": comparison,
                    "runbook": runbook,
                    "query": query,
                    "progress": [],
                }
                save(path, journal)
                session = api.sessions.create(runbook)
                journal.update(state="uncertain", session_id=session.session_id)
                save(path, journal)
                result = None
                try:
                    for event in api.sessions.turn_stream(
                        session.session_id, query=query, complete=True
                    ):
                        if isinstance(event, TurnResult):
                            result = event
                        else:
                            journal["progress"].append(event.model_dump(mode="json"))
                            save(path, journal)
                except Exception as error:
                    journal["error_category"] = type(error).__name__
                    save(path, journal)
                    raise
                if result is None:
                    raise RuntimeError("Stream ended without a completion")
                if crash:
                    os._exit(71)
            journal.update(state="response_saved", result=result.model_dump(mode="json"))
            save(path, journal)
            packet = export(work, journal, result, config)
            journal["state"] = packet["analysis_status"]
            save(path, journal)
            return packet

File: src/quality-investigation/src/main/java/io/ioka/demo/quality/Workflow.java
// SPDX-License-Identifier: Apache-2.0
package io.ioka.demo.quality;
import com.fasterxml.jackson.databind.JsonNode;
import com.fasterxml.jackson.databind.node.ObjectNode;
import io.ioka.munarium.client.model.*;
import io.ioka.munarium.client.model.SessionsApi.*;
import io.ioka.munarium.client.planes.Params;
import java.nio.file.*;
import java.util.*;
import java.util.regex.Pattern;

@SuppressWarnings("try")
public final class Workflow {
    private Workflow() {}
    public static JsonNode packet(Path work,String id,String revision,boolean recover,boolean crash,String endpoint) throws Exception {
        try(var lease=new Lease(work)) {
            Fixtures.verify(Path.of("/inputs"));Fixtures.require(Set.of("baseline","corrected").contains(revision),"Explicit baseline or corrected revision required");
            var grant=Bootstrap.grant();String name=Objects.requireNonNull(grant.runbooks().get(id));var record=Bootstrap.registry().path(name).path(revision);Fixtures.require(record.has("version"),"Selected revision has not been created");
            String key=FilesUtil.hash(grant.namespace()+grant.uid()+FilesUtil.json(record)+Files.readString(Path.of("/inputs/manifest.json")));
            Path journalPath=work.resolve("journal.json");ObjectNode journal;
            try(var reader=Bootstrap.reader();var api=Bootstrap.client(grant.token(),grant.uid(),endpoint)) {
                Bootstrap.assertFrozen(reader,record);
                if(Files.exists(journalPath)) {
                    journal=(ObjectNode)FilesUtil.read(journalPath);Fixtures.require(journal.path("key").asText().equals(key),"Changed investigation binding requires new work");
                    if(!journal.has("result")) {
                        Fixtures.require(recover && journal.has("session_id"),"Uncertain turn requires explicit transcript recovery; no replay");var session=api.sessions.get(journal.path("session_id").asText());
                        Fixtures.require(session.uid().equals(grant.uid()) && session.runbookRef().equals(record.path("runbook").asText()),"Transcript scope mismatch");
                        var turns=session.turns().stream().filter(t->t.query().equals(journal.path("query").asText()) && t.completion()!=null && !t.completion().isNull()).toList();Fixtures.require(turns.size()==1,"No uniquely recoverable completion");
                        var turn=turns.getFirst();var completion=(ObjectNode)turn.completion().deepCopy();completion.put("was_override",completion.path("resolved").path("was_override").asBoolean());
                        var result=new TurnResult(session.sessionId(),turn.ordinal(),turn.collectionsSearched(),null,Arrays.asList(Json.MAPPER.treeToValue(turn.hits(),TurnHit[].class)),Arrays.asList(Json.MAPPER.treeToValue(turn.envelope(),CollectionEnvelope[].class)),Json.MAPPER.treeToValue(completion,TurnCompletion.class),null);
                        journal.set("result",Json.MAPPER.valueToTree(result));journal.set("transcript_completion",turn.completion());journal.put("recovered",true);FilesUtil.save(journalPath,journal);
                    }
                } else {
                    journal=Json.MAPPER.createObjectNode();journal.put("key",key).put("query","Assemble the sampled inspection, narrative and required procedure for fictional "+id+". Preserve missing values and disagreement.").put("state","creating_session");journal.set("binding",record);FilesUtil.save(journalPath,journal);
                    journal.put("session_id",api.sessions.create(record.path("runbook").asText()).sessionId()).put("state","uncertain");FilesUtil.save(journalPath,journal);
                    var progress=journal.putArray("progress");
                    var result=api.sessions.turnStream(journal.path("session_id").asText(),Params.TurnOptions.of(journal.path("query").asText()).withCompletion(null),event->{progress.add(Json.MAPPER.valueToTree(event));try {FilesUtil.save(journalPath,journal);}catch(Exception e) {throw new IllegalStateException(e);}});
                    if(crash) Runtime.getRuntime().halt(71);journal.set("result",Json.MAPPER.valueToTree(result));FilesUtil.save(journalPath,journal);
                }
                Bootstrap.assertFrozen(reader,record);var facts=reader.query.facts(record.path("version").asText(),Params.FactsQuery.atSeq(record.path("pin").asLong()));
                var result=Json.MAPPER.treeToValue(journal.path("result"),TurnResult.class);var output=Json.MAPPER.createObjectNode();output.put("case_id",id).put("revision",revision).put("fictional",true).put("client_revision",Bootstrap.REVISION);output.set("binding",record);output.set("observations",Json.MAPPER.valueToTree(facts));output.set("evidence",journal.path("result"));output.put("recovered",journal.path("recovered").asBoolean());
                var errors=output.putArray("unresolved");ObjectNode answer=Json.MAPPER.createObjectNode();
                try {answer=validate(result,grant,name,id,facts);}
                catch(Exception e) {errors.add("Model evidence validation failed: "+e.getMessage());}
                if(result.hierarchy()==null) errors.add("Hierarchy decision is unavailable in the recovered transcript; completeness remains unresolved");
                else {
                    var layers=result.hierarchy().layers();if(layers.size()!=2 || !layers.get(0).layer().equals("observations") || !layers.get(0).block().equals("fact_slice") || !layers.get(1).block().equals("document_hits")) errors.add("Required fact or procedure evidence unavailable");
                }
                var observed=facts.facts().stream().filter(f->f.key().equals("defects")).findFirst();Integer defects=observed.map(f->Integer.valueOf(f.value())).orElse(null);Integer narrative=narrative(result.hits());
                if(defects==null) errors.add("Required inspection defect observation is missing");if(narrative==null) errors.add("Required narrative observation is missing");
                boolean conflict=defects!=null && narrative!=null && !defects.equals(narrative);output.put("disagreement",conflict);output.set("model_suggestion",answer);output.put("status",!errors.isEmpty()?"incomplete":conflict?"disagreement_requires_review":"complete_for_review");output.put("complete_lot_defect_count_available",false);
                output.put("supported_conclusion",conflict?"Recorded sample and narrative disagree; reconciliation is required":"Available observations are listed for review; this is not a lot-wide count");output.put("hypothesis","Root cause is not determined; further investigation is required");output.put("disposition","Reviewer decision required");
                FilesUtil.save(work.resolve("packet.json"),output);
                String text="# Fictional quality investigation | "+id+" | "+revision+"\n\nStatus: "+output.path("status").asText()+"\nLedger: "+record.path("version").asText()+" at pin "+record.path("pin").asLong()+"\nRunbook: "+record.path("runbook").asText()+"\n\nObservation: sampled defects = "+defects+"; narrative defects = "+narrative+"\nSupported conclusion: "+output.path("supported_conclusion").asText()+"\nHypothesis: "+output.path("hypothesis").asText()+"\nProcedure: "+answer.path("action").asText("unavailable")+"\nDisposition: reviewer decision required\nExact lot defect count: unavailable; procedure search cannot establish it\n\nCitations: "+answer.path("citations")+"\nUnresolved: "+errors+"\nSession: "+result.sessionId()+"\n";
                FilesUtil.write(work.resolve("packet.md"),text);journal.put("state",output.path("status").asText());FilesUtil.save(journalPath,journal);return output;
            }
        }
    }
    public static Integer narrative(List<TurnHit> hits) {for(var hit:hits) {var m=Pattern.compile("Narrative defects: ([0-9]+)").matcher(hit.text());if(m.find()) return Integer.valueOf(m.group(1));}return null;}
    public static ObjectNode validate(TurnResult result,Bootstrap.Grant grant,String collection,String id,Ledger.FactsPage facts) throws Exception {
        var c=result.completion();Fixtures.require(c!=null && c.provider().equals(grant.provider().equals("fixture")?"ollama":grant.provider()) && c.model().equals(grant.model()),"Unexpected model identity");Fixtures.require(c.verification()==null || c.verification().violations().isEmpty(),"Unresolved verification violations");
        String raw=c.text().trim().replaceFirst("^```(?:json)?\\s*","").replaceFirst("\\s*```$","");var parsed=Json.MAPPER.readTree(raw);Fixtures.require(parsed.isObject(),"Expected an object");var answer=(ObjectNode)parsed;
        var fields=new HashSet<String>();answer.fieldNames().forEachRemaining(fields::add);Fixtures.require(fields.equals(Set.of("observed_defects","narrative_defects","action","root_cause","hypothesis","citations")),"Unexpected answer schema");
        var expected=facts.facts().stream().filter(f->f.key().equals("defects")).findFirst();if(expected.isPresent()) Fixtures.require(answer.path("observed_defects").isIntegralNumber() && answer.path("observed_defects").asInt()==Integer.parseInt(expected.get().value()),"Inspection observation mismatch");else Fixtures.require(answer.path("observed_defects").isNull(),"Missing inspection was invented");
        Integer narrative=narrative(result.hits());Fixtures.require(narrative==null?answer.path("narrative_defects").isNull():answer.path("narrative_defects").isIntegralNumber() && answer.path("narrative_defects").asInt()==narrative,"Narrative observation mismatch");
        Fixtures.require(answer.path("root_cause").asText().equals("not determined") && answer.path("hypothesis").asText().equals("requires investigation") && answer.path("action").asText().equals("Hold the lot for quality review."),"Unsupported conclusion or disposition");
        var labels=new HashMap<String,TurnHit>();for(var hit:result.hits()) {Fixtures.require(hit.collection().equals(collection) && Set.of(collection+"/procedure.txt",collection+"/note.txt").contains(hit.sourcePath()),"Unexpected evidence scope");String file=hit.sourcePath().substring(collection.length()+1);Fixtures.require(hit.sourceContentHash().equals(FilesUtil.hash(Files.readString(Path.of("/inputs/"+id+"/"+file)))),"Source hash mismatch");labels.put("procedures/"+hit.chunkId(),hit);}
        Fixtures.require(answer.path("citations").isArray() && !answer.path("citations").isEmpty(),"Citations missing");var paths=new HashSet<String>();for(var citation:answer.path("citations")) {Fixtures.require(labels.containsKey(citation.asText()),"Unserved citation");paths.add(labels.get(citation.asText()).sourcePath());}Fixtures.require(paths.equals(Set.of(collection+"/procedure.txt",collection+"/note.txt")),"Both required document sources must be cited");return answer;
    }
}
